src/py/pyodide/console.py

```diff
@@ -158,57 +168,116 @@ def __call__(self, source, filename="<console>", symbol="single") -> CodeRunner:
         return super().__call__(source, filename, symbol)  # type: ignore
 
 
-class _InteractiveConsole(code.InteractiveConsole):
+RunCodeResult = Union[Tuple[Literal["exception"], str], Tuple[Literal["success"], Any]]
+RunSourceResult = Union[
+    Tuple[Literal["incomplete"], None],
+    Tuple[Literal["syntax-error"], str],
+    Tuple[Literal["complete"], Awaitable[RunCodeResult]],
+]
+
+INCOMPLETE: Literal["incomplete"] = "incomplete"
+SYNTAX_ERROR: Literal["syntax-error"] = "syntax-error"
+COMPLETE: Literal["complete"] = "complete"
+
+SUCCESS: Literal["success"] = "success"
+EXCEPTION: Literal["exception"] = "exception"
+
+
+class Console:
     """Interactive Pyodide console
 
-    Base implementation for an interactive console that manages
-    stdout/stderr redirection. Since packages are loaded before running
-    code, :any:`_InteractiveConsole.runcode` returns a JS promise.
+    An interactive console based on the Python standard library
+    `code.InteractiveConsole` that manages stream redirections and asynchronous
+    execution of the code.
 
-    ``self.stdout_callback`` and ``self.stderr_callback`` can be overloaded.
+    The stream callbacks can be modified directly as long as
+    `persistent_stream_redirection` isn't in effect.
 
     Parameters
     ----------
-    locals
-        Namespace to evaluate code.
-    stdout_callback
-        Function to call at each ``sys.stdout`` flush.
-    stderr_callback
-        Function to call at each ``sys.stderr`` flush.
-    persistent_stream_redirection
-        Whether or not the std redirection should be kept between calls to
-        ``runcode``.
+    globals : ``dict``
+
+        The global namespace in which to evaluate the code. Defaults to a new empty dictionary.
```

Comment:
I think I made these based on looking at the sphinx output. Without the newlines the docs don't end up looking correct. I don't know why, but I think it's important to check what the output actually looks like rather than trusting the spec. (I could be wrong about this though, I will double check.)

Reply:
Current docstring:
```
    globals : ``dict``

        The global namespace in which to evaluate the code. Defaults to a new empty dictionary.

    stdout_callback : Callable[[str], None] Function to call at each write to
        ``sys.stdout``. Defaults to ``None``.
```
Result:
![image](https://user-images.githubusercontent.com/8739626/126239729-b3ef7473-d61c-416f-aa72-97e0ef5d4a0f.png)

With newline removed:
```
    globals : ``dict`` The global namespace in which to evaluate the code. Defaults to a new empty dictionary.
```

![image](https://user-images.githubusercontent.com/8739626/126239643-3d2ce306-5a01-43d4-a91a-d042d84313d2.png)

It looks fine when input has no empty lines:
```
    globals : ``dict`` 
        The global namespace in which to evaluate the code. Defaults to a new empty dictionary.

    stdout_callback : Callable[[str], None] 
        Function to call at each write to ``sys.stdout``. Defaults to ``None``.
```

Maybe that's what you meant?

Reply:
Okay that's clearly what you meant.